### PR TITLE
CiliumEndpointBatch support in cilium-agent

### PR DESCRIPTION
**Pull request by krishgobinath**
1) Implement CiliumEndpointBatch controller
2) Convert CEB update into multiple CEP updates
3) Call endpointUpdate and endpointDelete functions for changes in CEP
4) If CEB feature is enabled, run CEB controller instead of CEP
controller

Signed-off-by: Gobinath Krishnamoorthy gobinathk@google.com

### CODE REVIEW

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		nodeIP = node.GetIPv6().String()
+	}
+
+	for i, ep := range ceb.Endpoints {
+		log.WithFields(logrus.Fields{
+			"CebName":          ceb.GetName(),
+			"CoreEndpointName": ep.Name,
+		}).Debug("New CEB added, Calling CoreEndpointUpdate")
+		c := k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint(&ceb.Endpoints[i])
+		// Skip processing endpointupdate for local node CEPs
```
Comment (aanm) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: If the node IP changes between cilium-agent restarts, we will receive all CEPs from the previous cilium instance and we will execute `k.endpointUpdated(nil, c)` for those endpoints. Is this safe? I assume it isn't given that this comment is stating that we are skipping the process of local node CEPs.

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+// 3) any existing core CEPs are modified in CEB
+// call endpointUpdated/endpointDeleted only for remote node CEPs.
+func (k *K8sWatcher) endpointBatchUpdated(oldCeb, newCeb *cilium_v2a1.CiliumEndpointBatch) {
+
+	nodeIP := node.GetIPv4().String()
+	if option.Config.EnableIPv6 {
+		nodeIP = node.GetIPv6().String()
+	}
+
+	oldCeps := make(map[string]*cilium_v2a1.CoreCiliumEndpoint)
```
Comment (aanm) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: nit, the `oldCeps` and `newCeps` can be initialized with length of `len(oldCeb.Endpoints)` and `len(newCeb.Endpoints)` respectively.

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+			}
+			k.endpointDeleted(c)
+		}
+	}
+
+	// Process if any new items added into newCeb
+	for epName, newEp := range newCeps {
+		if _, exists := oldCeps[epName]; !exists {
+			log.WithFields(logrus.Fields{
+				"CebName":          oldCeb.GetName(),
```
Comment (aanm) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: this should probably be `newCeb.GetName()`?

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	// process if any cep entries changed from old to new
+	for epName, newCep := range newCeps {
+		if oldCep, exists := oldCeps[epName]; exists {
+			if oldCep.DeepEqual(newCep) {
+				continue
+			}
+			log.WithFields(logrus.Fields{
+				"CebName":          oldCeb.GetName(),
```
Comment (aanm) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: this should probably be `newCeb.GetName()`?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 1 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: The primary reason for skipping some `CEP`  is that `localNode` already has the latest and greatest CEP, so we don't want any updates for the `localNode`. However, only on the `remoteNodes` we want to update the `CEP`. 

In the scenario you had mentioned, there is nothing wrong in calling `k.endpointUpdated(nil, c)` for the localNode CEPs, this isn't a destructive operation. Eventually `k.endpointUpdated(nil, c)` would be called for updated CEP.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 2 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: SG. Done.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 3 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: why do you think this should be `newCeb.GetName()` ? 
Both `oldCeb.GetName()` and `newCeb.GetName()` should return same value inside this function `endpointBatchUpdated`

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 4 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: resolving this conversation, following up on previous comment.

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	nodeIP := node.GetIPv4().String()
+	if option.Config.EnableIPv6 {
+		nodeIP = node.GetIPv6().String()
+	}
+
+	for i, ep := range ceb.Endpoints {
+		log.WithFields(logrus.Fields{
+			"CebName":          ceb.GetName(),
+			"CoreEndpointName": ep.Name,
+		}).Debug("New CEB added, Calling CoreEndpointUpdate")
```
Comment (Weil0ng) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: nit: s/Calling/calling/g

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 1 above)
Comment (Weil0ng) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Instead of checking nodeIP, can we check if `CEP.ID` is in local endpointManager's store[1]?

[1] https://github.com/cilium/cilium/blob/master/pkg/endpointmanager/manager.go#L49

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	// process removed entries from ceb. old ceb would have one or more stale cep
+	// entries. new ceb is source of truth.
+	for epName, oldEp := range oldCeps {
+		if _, exists := newCeps[epName]; !exists {
+			log.WithFields(logrus.Fields{
+				"CebName":          oldCeb.GetName(),
+				"CoreEndpointName": epName,
+			}).Debug("CEP deleted, Calling endpointDeleted")
```
Comment (Weil0ng) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: nit: s/Calling/calling/g

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+		}
+	}
+
+	// Process if any new items added into newCeb
+	for epName, newEp := range newCeps {
+		if _, exists := oldCeps[epName]; !exists {
+			log.WithFields(logrus.Fields{
+				"CebName":          oldCeb.GetName(),
+				"CoreEndpointName": epName,
+			}).Debug("CEB updated, Calling endpointUpdated")
```
Comment (Weil0ng) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: nit: s/Calling/calling/g

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+	// process if any cep entries changed from old to new
+	for epName, newCep := range newCeps {
+		if oldCep, exists := oldCeps[epName]; exists {
+			if oldCep.DeepEqual(newCep) {
+				continue
+			}
+			log.WithFields(logrus.Fields{
+				"CebName":          oldCeb.GetName(),
+				"CoreEndpointName": epName,
+			}).Debug("CEB updated, Calling endpointUpdated")
```
Comment (Weil0ng) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: nit: s/Calling/calling/g

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 5 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Done.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 1 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Updated code, by using `LookupPodName` function to check `cep` is in localNode.

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+// determines below things
+// 1) any new core CEPs are packed in CEB
+// 2) any core CEPs are removed from CEB
+// 3) any existing core CEPs are modified in CEB
+// call endpointUpdated/endpointDeleted only for remote node CEPs.
+func (k *K8sWatcher) endpointBatchUpdated(oldCeb, newCeb *cilium_v2a1.CiliumEndpointBatch) {
+
+	oldCeps := make(map[string]*cilium_v2a1.CoreCiliumEndpoint, len(oldCeb.Endpoints))
+	for i, ep := range oldCeb.Endpoints {
+		oldCeps[ep.Namespace+ep.Name] = &oldCeb.Endpoints[i]
```
Comment (Weil0ng) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: should this be `ep.Namespace + '/' + ep.Name`?

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	newCeps := make(map[string]*cilium_v2a1.CoreCiliumEndpoint, len(newCeb.Endpoints))
+	for i, ep := range newCeb.Endpoints {
+		newCeps[ep.Namespace+ep.Name] = &newCeb.Endpoints[i]
```
Comment (Weil0ng) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: ditto

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 9 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Functionally It wouldn't make any difference, `oldCeps` and `newCeps` are temporary maps to hold unique ceps.

Comment (Weil0ng) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: https://github.com/cilium/cilium/pull/16945#discussion_r673874610

Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Thanks for reminding. Done.

**Review by Weil0ng — APPROVED** [review]

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/sirupsen/logrus"
+)
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: We order imports as follows

```
stdlib

cilium packages

3rd party
```

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+
+func (k *K8sWatcher) ciliumEndpointBatchInit(client *k8s.K8sCiliumClient, asyncControllers *sync.WaitGroup) {
+
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: Nit: we don't need a newline at the start of new functions

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+						}
+						k.endpointBatchUpdated(oldCeb, newCeb)
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if ceb := k8s.ObjToCiliumEndpointBatch(obj); ceb != nil {
+					k.endpointBatchDeleted(ceb)
+				}
+			},
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: For new objects, we want to make sure we're using the newly refactored the event handling from k8s watchers via the `subscriber` package as per https://github.com/cilium/cilium/issues/15471. 

What I'm proposing is to avoid defining methods on the `K8sWatcher` that's specific for CEBs, such as `endpointBatchAdded` et al and instead create a specific "subscriber" for CEBs. See https://github.com/cilium/cilium/pull/15295 as well

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 1 above)
Comment (christarazi) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: > The primary reason for skipping some CEP is that localNode already has the latest and greatest CEP, so we don't want any updates for the localNode. However, only on the remoteNodes we want to update the CEP.

Can we add that to the comment in the code?

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	// process removed entries from ceb. old ceb would have one or more stale cep
+	// entries. new ceb is source of truth.
+	for epName, oldEp := range oldCeps {
+		if _, exists := newCeps[epName]; !exists {
+			log.WithFields(logrus.Fields{
+				"CebName":          oldCeb.GetName(),
+				"CoreEndpointName": epName,
+			}).Debug("CEP deleted, calling endpointDeleted")
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: Style nit: for the sake of consistency in the code, for abbreviations such as CEP or CEB, let's keep it as one style--all upper-cased, including for text in log msgs or comments and in variable names. https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if ceb := k8s.ObjToCiliumEndpointBatch(obj); ceb != nil {
+					k.endpointBatchDeleted(ceb)
+				}
+			},
+		},
+		nil,
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: Minor: if you look at other k8s watcher event handler functions, most have the conversion function set here. In this case it would be `k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint` upfront. This avoids having to call it explicitly in the handler and instead it passes that result on to the handler (`endpointBatchDeleted`, et al.).

**Review by christarazi — CHANGES_REQUESTED** [review]

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 15 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Your suggestion may simplify code, But I don't see much beneficial overall. If you look at the code, we are doing this conversion`k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint` only for  required  `CEPs` not for all `CEPs`.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 13 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Done.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 14 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Done.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 11 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Done.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 1 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Added comment.

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+var (
+	cebSubscriber = subscriber.NewRaw()
+)
+
+func (k *K8sWatcher) ciliumEndpointBatchInit(client *k8s.K8sCiliumClient, asyncControllers *sync.WaitGroup) {
+	log.Info("Initializing CEB controller")
+	var once sync.Once
+
+	// Register for all ceb updates.
+	cebSubscriber.Register(newCEBSubscriber(k))
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: Rather than dep inject the entire `K8sWatcher`, let's instead only inject individual isolated deps. Otherwise, the purpose of the refactor is rendered moot.

Comment (christarazi) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: On second thought, breaking up the `K8sWatcher` dependency will depend on converting the CEP handling into a separate subscriber[1] like you've done here for CEB. If done, then this refactor should be done in its own commit for the sake of bisect. You can either do this refactor now as part of this PR, or do it as a followup as another PR into the long-running branch. Either way it will need to be done because we want all new additions to the K8s watchers to be using the new scheme and over time, all other watchers will get converted.

File: pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+			"CEBName": ceb.GetName(),
+			"CEPName": ep.Name,
+		}).Debug("CEB added, calling CoreEndpointUpdate")
+		c := k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint(&ceb.Endpoints[i])
+		// LocalNode already has the latest CEP.
+		// Hence, skip processing endpointupdate for localNode CEPs.
+		if p := cs.kWatcher.endpointManager.LookupPodName(k8sUtils.GetObjNamespaceName(c)); p != nil {
+			continue
+		}
+		cs.kWatcher.endpointUpdated(nil, c)
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go:49]: [1]

File: pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+func newCEBSubscriber(k *K8sWatcher) *CEBSubcriber {
+	return &CEBSubcriber{
+		kWatcher: k,
+	}
+}
+
+func (cs *CEBSubcriber) OnAdd(obj interface{}) {
+	ceb, _ := obj.(*cilium_v2a1.CiliumEndpointBatch)
+	cs.CEBAdded(ceb)
+}
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go]: A new subscriber should be in it's own package rather than as part of the K8s watcher (that is the point of the refactor). If there already exists a package for CEB, then this code belongs there. 

Additionally, we can define our own "subscriber" by defining a new method set (inside `pkg/k8s/watchers/subscriber/ceb.go` for ex.) for `CiliumEndpointBatch` objects, rather than using the "raw" subscriber that forces us to type assert due to the methods accepting only `interface{}`. See `pkg/k8s/watchers/subscriber/service.go`.

File: pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+
+import (
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	k8sUtils "github.com/cilium/cilium/pkg/k8s/utils"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CEBSubcriber struct {
```
Comment (christarazi) [inline pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go]: Nit: consider only exporting objects / methods only if they're referenced outside the package. This seems like it should just be `cebSubscriber`.

**Review by christarazi — CHANGES_REQUESTED** [review]

File: pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go (hunk 19 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go]: SG Done.

**Review by christarazi — CHANGES_REQUESTED** [review]
If the metrics commit has been reverted, then we can drop the 2 commits regarding it as well.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 16 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: `CEB` relies on the `CEP` dependencies. Breaking CEP dependencies should be handled in another PR, not in this PR.

File: pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go (hunk 18 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go]: Cilium-agent doesn't use the actual `CEB` package, it watches for the `CEB`s and calls the `CEP` handler. There is nothing more we do in `cilium-agent`. The actual CEB logic resides in the operator. hence, this code is part of `k8swatcher`.

Main reason I had skipped having a separate `ceb subscriber` is it will be a replica of `pkg/k8s/watchers/subscriber/raw.go`.I prefer having lesser code, rather than having duplicate code.

File: pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go (hunk 17 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go:49]: resolving based on previous comment.

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 16 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Tracking the request change in GH [Issue](https://github.com/cilium/cilium/issues/17355)

File: pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go (hunk 18 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go]: Updated PR with separate CEB subscriber.

File: pkg/k8s/watchers/subscriber/ceb.go [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+// limitations under the License.
+
+package subscriber
+
+import (
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+)
+
+// NewCEB creates a new ceb subscriber list.
+func NewCEB() *CebList {
```
Comment (christarazi) [inline pkg/k8s/watchers/subscriber/ceb.go]:
```suggestion
func NewCEB() *CEBList {
```

Please `s/Ceb/CEB/g` across this file

**Review by christarazi — APPROVED** [review]
LGTM thanks. Small, non-blocking comment on consistency of capitalization

File: pkg/k8s/watchers/subscriber/ceb.go (hunk 20 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/subscriber/ceb.go]: Done.

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+)
+
+func (k *K8sWatcher) ciliumEndpointBatchInit(client *k8s.K8sCiliumClient, asyncControllers *sync.WaitGroup) {
+	log.Info("Initializing CEB controller")
+	var once sync.Once
+
+	// Register for all ceb updates.
+	cebNotify.Register(newCEBSubscriber(k))
+
+	_, cebController := informer.NewInformer(
```
Comment (errordeveloper) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]: I think it's best to rename this to make it clearer that it's not the controller itself.
```suggestion
	_, cebInformer := informer.NewInformer(
```

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+	)
+	once.Do(func() {
+		// Signalize that we have put node controller in the wait group
+		// to sync resources.
+		asyncControllers.Done()
+	})
+	k.blockWaitGroupToSyncResources(
+		wait.NeverStop,
+		nil,
+		cebController.HasSynced,
```
Comment (errordeveloper) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]:
```suggestion
		cebInformer.HasSynced,
```

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+		k8sAPIGroupCiliumEndpointBatchV2Alpha1,
+	)
+
+	go cebController.Run(wait.NeverStop)
```
Comment (errordeveloper) [inline pkg/k8s/watchers/cilium_endpoint_batch.go]:
```suggestion
	go cebInformer.Run(wait.NeverStop)
```

File: pkg/k8s/watchers/subscriber/ceb.go [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+// NewCEB creates a new ceb subscriber list.
+func NewCEB() *CEBList {
+	return &CEBList{}
+}
+
+// CebHandler is implemented by event handlers responding to
+// CiliumEndpointBatch events.
+type CebHandler interface {
```
Comment (errordeveloper) [inline pkg/k8s/watchers/subscriber/ceb.go]:
```suggestion
type CEBHandler interface {
```

File: pkg/k8s/watchers/subscriber/ceb.go [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+	OnAdd(ceb *cilium_v2a1.CiliumEndpointBatch)
+	OnUpdate(oldCeb, newCeb *cilium_v2a1.CiliumEndpointBatch)
+	OnDelete(ceb *cilium_v2a1.CiliumEndpointBatch)
+}
+
+// Register registers the CEB event handler as a subscriber.
+func (l *CEBList) Register(c CebHandler) {
```
Comment (errordeveloper) [inline pkg/k8s/watchers/subscriber/ceb.go]:
```suggestion
func (l *CEBList) Register(c CEBHandler) {
```

File: pkg/k8s/watchers/subscriber/ceb.go [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+	for _, s := range l.subs {
+		s.OnDelete(ceb)
+	}
+}
+
+// CEBList holds the CEB subscribers to any CiliumEndpointBatch resource / object changes in
+// the K8s watchers.
+type CEBList struct {
+	list
+	subs []CebHandler
```
Comment (errordeveloper) [inline pkg/k8s/watchers/subscriber/ceb.go]:
```suggestion
	subs []CEBHandler
```

File: pkg/k8s/watchers/subscriber/service.go [hunk 27]
```diff
@@ -57,7 +57,7 @@ func (l *ServiceList) NotifyUpdate(oldSvc, newSvc *slim_corev1.Service) {
 	}
 }
 
-// NotifyDelete notifies all the subscribers of an update event to a service.
+// NotifyDelete notifies all the subscribers of a delete event to a service.
```
Comment (errordeveloper) [inline pkg/k8s/watchers/subscriber/service.go:60]: Should this and the other comment fix go in a separate commit?

File: operator/main.go [hunk 28]
```diff
@@ -369,7 +369,7 @@ func onOperatorStartLeading(ctx context.Context) {
 		// Start CEP watcher
 		operatorWatchers.CiliumEndpointsBatchInit(k8s.CiliumClient().CiliumV2(), cebController)
 		// Start the CEB controller, after current CEPs are synced locally in cache.
-		cebController.Run(operatorWatchers.CiliumEndpointStore, stopCh)
+		go cebController.Run(operatorWatchers.CiliumEndpointStore, stopCh)
```
Comment (errordeveloper) [inline operator/main.go:372]: This looks like a bug fix, a separate commit might make more sense.

File: pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go [hunk 29]
```diff
@@ -0,0 +1,143 @@
...
+
+// OnAdd invoked for newly created CEBs, iterates over coreCEPs
+// packed in the CEB, converts coreCEP into types.CEP and calls endpointUpdated only for remoteNode CEPs.
+func (cs *cebSubscriber) OnAdd(ceb *cilium_v2a1.CiliumEndpointBatch) {
+	for i, ep := range ceb.Endpoints {
+		log.WithFields(logrus.Fields{
+			"CEBName": ceb.GetName(),
+			"CEPName": ep.Name,
+		}).Debug("CEB added, calling CoreEndpointUpdate")
+		c := k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint(&ceb.Endpoints[i])
```
Comment (errordeveloper) [inline pkg/k8s/watchers/cilium_endpoint_batch_subscriber.go:43]: Should probably check the length of the slice...

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 21 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go]: Done.

File: pkg/k8s/watchers/subscriber/ceb.go (hunk 25 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/subscriber/ceb.go]: Done.

File: operator/main.go (hunk 28 above)
Comment (krishgobinath) [reply operator/main.go:372]: Thats right, this is bug fix. Whole feature isn't in ready state,. Hopefully with this PR, CiliumEndpointBatch feature will be in working state. Going forward we can think about creating new PR for bug fixes.

File: pkg/k8s/watchers/subscriber/service.go (hunk 27 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/subscriber/service.go:60]: If there is code change surely Yes. Since its minor comment change, i think it's okay.

File: pkg/k8s/watchers/cilium_endpoint_batch.go [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	cebNotify = subscriber.NewCEB()
+)
+
+func (k *K8sWatcher) ciliumEndpointBatchInit(client *k8s.K8sCiliumClient, asyncControllers *sync.WaitGroup) {
```
Comment (aanm) [inline pkg/k8s/watchers/cilium_endpoint_batch.go:35]: This function should be similar to the `ciliumEndpointsInit`, i.e., we don't need to keep watching for `ciliumEndpointBatch` if the kvstore is enabled.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 30 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go:35]: Done.

Comment (aanm) [reply pkg/k8s/watchers/cilium_endpoint_batch.go:35]: Unresolved.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/watchers/cilium_endpoint_batch.go (hunk 30 above)
Comment (krishgobinath) [reply pkg/k8s/watchers/cilium_endpoint_batch.go:35]: Updated to reflect code in `ciliumEndpointsInit`.

**Review by aanm — APPROVED** [review]
